lib/spack/spack/solver/asp.py

```diff
@@ -4582,14 +4516,38 @@ def __init__(self):
         self.selector = ReusableSpecsSelector(configuration=spack.config.CONFIG)
 
     @staticmethod
-    def _check_input_and_extract_concrete_specs(specs):
-        reusable = []
+    def _check_input_and_extract_concrete_specs(
+        specs: List[spack.spec.Spec],
+    ) -> List[spack.spec.Spec]:
+        reusable: List[spack.spec.Spec] = []
+        analyzer = create_graph_analyzer()
         for root in specs:
             for s in root.traverse():
                 if s.concrete:
                     reusable.append(s)
-                elif spack.repo.PATH.is_virtual(s.name):
-                    continue
+                else:
+                    if spack.repo.PATH.is_virtual(s.name):
+                        continue
+                    # Error if direct dependencies cannot be satisfied
+                    deps = {edge.spec.name for edge in s.edges_to_dependencies() if edge.direct}
```

Comment:
when does this come up?

Reply:
I expect this to come up mainly for packages that don't need a compiler:
```console
$ spack spec py-cinemasci %gcc
==> Error: 'py-cinemasci %gcc' cannot depend on gcc
```